text/0000-machine-input.md

```diff
@@ -0,0 +1,183 @@
+- Machine is dynamically created; cannot guarantee that the machine is the same shape every time (even though it should always be)
+- Machine definition cannot be serialized; must be run ad-hoc rather than being directly interpreted
+- Machine is no longer easily portable between different languages (future goal)
+- No standard way of creating a machine factory
+
+**Bespoke input event**
+
+```js
+const greetMachine = createMachine({
+  on: {
+    'GREET.INPUT': {
+      actions: (_, event) => { console.log(`Hello, ${event.name}!`); }
+    }
+  }
+});
+
+const greetService = interpret(greetMachine).start();
+
+greetService.send({ type: 'GREET.INPUT', name: 'David' });
+```
+
+Problems:
+- Since this input event is bespoke, there is no built-in way of identifying input properties
+- Input event _must_ be first event sent; this is not guaranteed
+- Cannot use input data at semantic start of machine; e.g., not usable in `entry`
+- No standard input event shape
+
+-----
+
+**Prior art:**
+
+- [Data in AWS Step Functions (States Language)](https://states-language.net/spec.html#data):
+
+> The interpreter passes data between states to perform calculations or to dynamically control the state machine’s flow. All such data MUST be expressed in JSON.
+>
+> When a state machine is started, the caller can provide an initial JSON text as input, which is passed to the machine's start state as input. If no input is provided, the default is an empty JSON object, `{}`. As each state is executed, it receives a JSON text as input and can produce arbitrary output, which MUST be a JSON text. [...]
+
+- [Azure Logic Apps](https://docs.microsoft.com/en-us/azure/logic-apps/logic-apps-overview) have _triggers_, which are events (with input data) that start a logic apps workflow.
+- [Google Cloud Platform Workflows](https://cloud.google.com/workflows/docs/passing-runtime-arguments) enable `params` that let you specify data in a workflow execution request
+- [Pipedream](https://pipedream.com/docs/workflows/steps/triggers/#app-based-triggers) has triggers that can contain input data for the start of a workflow.
+- [Argo workflow inputs](https://argoproj.github.io/argo-workflows/workflow-inputs/)
+- [GitHub Actions inputs](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputs)
+
+## Detailed design
+
+**Overall API**
+
+This RFC proposes to extend `{ type: 'xstate.init' }`, which is always the first event that a machine receives, with the optional `data?: ...` property:
+
+```ts
+const greetingMachine = createMachine({
+  schema: {
+    // For type-safety; can be used with JSON Schema
+    input: {} as { name: string }
+  },
+  entry: (_, event) => {
+    // event.type is 'xstate.init'
+    // event.data is { name: ... }
+    console.log(`Hello, ${event.data.name}!`);
+  }
+});
+
+// Non-interpreted environments
+const greetingMachineWithInput = greetingMachine
+  .withInput({ name: 'David' });
```

Comment:
`input` seems to be rather **execution** specific and machines act mostly as "templates". If we'd extend the analogy to classic OOP then `input` looks mostly like a constructor parameter whereas the machine is the class itself.

WDYT about this?
```js
interpret(machine, { input }).start()
```

Reply:
I was also thinking about this in the past:
```ts
createMachine({
  context: initEvent => {
    return { /* construct initial context */ }
  }
})
```

One neat thing about this would be that the final return type of `context` factory would make for a complete, type-safe, `TContext`.

Reply:
The problem is that you could say the same thing about `context` itself (e.g. `machine.withContext(...)`). 

Practically, `machine.withInput(input)` is most useful in environments where a machine is _not_ to be interpreted, but rather tested. If we only had `interpret(machine, { input })`, there would be no opportunity to e.g. unit test the machine with a given input, since `machine.initialState` gives us no opportunity to provide the initial `xstate.init` event with `data`.

Reply:
> The problem is that you could say the same thing about context itself (e.g. machine.withContext(...)).

And I probably would say that now - with `input` in place (and other proposed changes). Previously we just didn't have good means/place to provide such per-execution arguments - it wasn't impossible or hard to introduce them but we just never stopped for a moment to think about alternative designs in this area.

Funny enough a bug for v5 has just been reported [here](https://github.com/statelyai/xstate/issues/3349) and the issue there is very much related to `withContext` and how you can now provide lazy functions at the construction time and through `provide` (and by extension through `withContext` and through `context` option in `useMachine`).

The problem I'm thinking about is that it's getting pretty complex (especially for the problem at hand) to "resolve" the final context value because stuff like this is valid:
```js
const m = createMachine({ context }).withContext(more).withContext(evenMore)
useMachine(m, { context: MOAR })
```
And each of those values can be a lazy function! It, of course, isn't impossible to keep all of them around and resolve them appropriately but I think that it shows how this is somewhat complicated.

If we took a step back and treated this as a "static" definition for context (and the only place to define it!) then things would get easier and clearer:
```js
createMachine({
  context: { some: 'value' }
})
```

But what about when you don't know the full context value upfront? Well, just use lazy context + input:
```js
createMachine({
  context: (initEvent) => ({ some: 'value', ...initEvent.data })
})
```

This is quite neat because we can easily guarantee that the context value ends up being "complete" and we simplify how many operations might be made to determine it. It's always just a combination of that single `config.context` and the `initEvent`. 

> since machine.initialState gives us no opportunity to provide the initial xstate.init event with data.

I see your point here but that's perhaps just a problem of the `machine.initialState` API in this case and not the proposed idea. After all we could just have smth like this:
```js
machine.getInitialState(input)
```

I'm not proposing this exact API (yet), just saying that it's not an unsolvable problem.

Reply:
Revised the APIs 👍